algorithms/scripts/BernoulliRBM-LogisticRegressionCV-yl565.py

```diff
@@ -0,0 +1,199 @@
+    """
+    scores = mad(x) 
+    return scores, np.array([np.NaN]*len(scores))
+
+
+def gen_pipeline(n_feat, n_RBM_cmp):
+    """Generate classification pipeline
+    Steps:
+        1. Select n_feat number of features with highest MAD.
+        2. Scale the n_feat features to within 0-1 range.
+        3. Fitting a Restricted Boltzmann Machine (RBM) with n_feat visible units and n_RBM_cmp hidden units.
+        4. Train a logistic regression classifier with the output of RBM as input features.
+    """
+    # Step 1: feature selection
+    if n_feat is None:
+        pipe=[]  # Skip feature selection, use all features
+    else:
+        pipe=[SelectKBest(fs_mad, k=n_feat)]
+    
+    if n_RBM_cmp is not None:
+        # Step 2: feature scaling
+        pipe.append(NormDist())  # Mapping input feature to [0, 1] range assuming columns are Guassian
+
+        # Step 3: Feature learning
+        pipe.append(
+            BernoulliRBM(random_state=0,
+                         n_components=n_RBM_cmp,
+                         verbose=True,
+                         learning_rate=0.01,
+                         n_iter=30))
+    
+    # Step 4: Classification
+    pipe.append(LogisticRegressionCV(class_weight='balanced', random_state=0))
+    
+    return make_pipeline(*tuple(pipe))
+
+
+def eval_model(fitted_model, X_train, X_test, y_train, y_test, title_str):
+    """Evaluate and plot the ROC curve of fitted_model
+    """
+    y_pred_train = fitted_model.decision_function(X_train)
+    y_pred_test = fitted_model.decision_function(X_test)
+
+    def get_threshold_metrics(y_true, y_pred):
+        roc_columns = ['fpr', 'tpr', 'threshold']
+        roc_items = zip(roc_columns, roc_curve(y_true, y_pred))
+        roc_df = pd.DataFrame.from_items(roc_items)
+        auroc = roc_auc_score(y_true, y_pred)
+        return {'auroc': auroc, 'roc_df': roc_df}
+
+    metrics_train = get_threshold_metrics(y_train, y_pred_train)
+    metrics_test = get_threshold_metrics(y_test, y_pred_test)
+
+    # Plot ROC
+    plt.figure()
+    for label, metrics in ('Training', metrics_train), ('Testing', metrics_test):
+        roc_df = metrics['roc_df']
+        plt.plot(roc_df.fpr, roc_df.tpr,
+            label='{} (AUROC = {:.1%})'.format(label, metrics['auroc']))
+    plt.xlim([0.0, 1.0])
+    plt.ylim([0.0, 1.05])
+    plt.xlabel('False Positive Rate')
+    plt.ylabel('True Positive Rate')
+    plt.title('Predicting TP53 mutation from gene expression (ROC curves)\n' + title_str)
+    plt.legend(loc='lower right')
+    plt.show()
+    return metrics_train['auroc'], metrics_test['auroc']
+
+
+# ## Fit model and show results
+
+# In[4]:
+
+# Fit the model (the computationally intensive part)
+n_features = [500, 2000, 5000]
+n_hiddens = [None, 200, 500, 1000]
+results={'n_feature': [], 'n_RBM_hidden': [], 'train_AUROC': [], 'test_AUROC': []}
+for n_f in n_features:
+    for n_h in n_hiddens:
+        if n_h is not None and n_f < n_h:
```

Comment:
I don't work with RBM very often either but I tend to think of RBM as a dimension reduction technique so I figured it's probably not worth producing more features than the raw ones especially consider our sample size. Could there be any case when it is beneficial for the RBM have more outputs than inputs? It would be very interesting.